### PR TITLE
Improve parsing performance with large collection expressions of numbers

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes [AB#1934136](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1934136)

The core issue here is that when the parser sees `[...` at the start of an expression it has to ask:

1. is this an attribute list that is going to be on a lambda? if so, parse as such.  If not:
2. parse as collection expression.

However, common collection-expr patterns can be very expensive for '1' currently.  Even simple, idiomatic code like `[1, 2, 3, 4]`.  This is because attribute parsing starts, and then immediatley goes off hte rails with error recovery as each element is not a valid attribute (and each is not hte start of some higher parsing construct it can pop out to).  As such, we get into an expensive loop, going through each element, and making all the diagnostics/annotations/skip work for all of them.  When the collections are large (thousands of elements is reasonable), this becomes a large problem.

The simple solution here is to just trivially detect that we're in a case taht is not, and should never be considered an attribute, and bail out immediately.  We then properly go into the collection-expr parsing loop without all that overhead.  Specifically if we see it's definitely not an attribute, then we use the heuristic of `[ <literal>` is decide it's definitely a collection expr.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Syntax/Parsing/CollectionExpressionParsingTests.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/CollectionExpressionParsingTests.cs:16864]: test was more so i could just trivially debug and validate the new behavior.

File: src/Compilers/CSharp/Test/Syntax/Parsing/LambdaParameterParsingTests.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/LambdaParameterParsingTests.cs:5664]: all below are small changes or improvements in error recovery cases.

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson @cston ptal.

File: src/Compilers/CSharp/Test/Syntax/Parsing/LambdaParameterParsingTests.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Syntax/Parsing/LambdaParameterParsingTests.cs:5669]: So is the `[` in "skipped tokens" now?

File: src/Compilers/CSharp/Test/Syntax/Parsing/CollectionExpressionParsingTests.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Syntax/Parsing/CollectionExpressionParsingTests.cs:16864]: Do we still have a problem when a large collection-expr syntactically looks like a valid attribute list, e.g. `[a, b, c, d, e, f, ...]`

Do we still have a problem when the start of the collection-expr looks like an attribute list? e.g. `[x, 2, 3, 4, 5, 6, ...]`

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Syntax/Parsing/CollectionExpressionParsingTests.cs:16864]: > Do we still have a problem when a large collection-expr syntactically looks like a valid attribute list, e.g. [a, b, c, d, e, f, ...]

No!  In that case, attribute parsing doesn't go 'off the rails'.  so we blitz through.  We do throw away rthe work.  Not ideal... but extremely hard to avoid.

> Do we still have a problem when the start of the collection-expr looks like an attribute list? e.g. [x, 2, 3, 4, 5, 6, ...]

Yes!  We would have the same issue then.  However, this PR wasn't an attempt to solve all the cases.  Just hte very realistic case of just numbers.  Basically: embedded binary blobs.  

We could def expand this further in the future though, esp. based on customer need.  THe idea then would be to make attribute parsing better about 'giving up' once things got bad enough.  But i think we can hold off on that for now.

**Comment by AlekseyTs** [comment]
@CyrusNajmabadi 
> The simple solution here is to just trivially detect that we're in a case taht is not, and should never be considered an attribute

Could you clearly state the exact rules used for that?

**Comment by CyrusNajmabadi** [comment]
> Could you clearly state the exact rules used for that?

Currently, if we see `[id` or `[target:` then we'll go the 'attribute' route.  If we have `[` and don't have that, then we make a choice:

1. First, note that this is a case where we now know it's not an attribute.  (Since those always must start with the things we checked for).
2. However, the old code always tried it as an attribute, so it's ok if we still fall back to that.  So for some common sorts of cases, we still do since we think the attribute fallback is still preferred.

So, the cases where we do this are:

```c#
[
namespace
```

more likely an incomplete attribute.

```c#
[
modifier
```

```c#
[
class
```

```c#
[
[
etc.
```

An incomplete attribute before a real attribute

If it's none of those, then we presume it's a collection-expr instead.

**Comment by CyrusNajmabadi** [comment]
Actually, i decided to go another, simpler, route.  Now it's simply that if we see `[ <literal>` that we say "this def isn't an attribute decl, treat this as a collection expr.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (ToddGrun) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:888]: >EatToken(); [](http://example.com/codeflow?start=12&length=11)

Is it normal for one of the Is* methods to modify state by eating a token?

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:888]: 100%.  THat's why there is a resetpoint.  You can speculatively parse to your heart's content (as long as you return back to the original parser state when the functino exits).

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (ToddGrun) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:896]: Are these two checks necessary? If not, is the thought that they make this code's intent more clear?

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:896]: the ISAttributeTarget check is *definitely* necessary.  As `true` can both be a literal, or an attribute-target (targets are any *word*, including keywords).  

The IsTrueIdentifier is technically *not* necessary.  But i prefer the clarity.  This means that the code says: if this is definitely an attribute, continue parsing as such.  If it is def not an attribute, then parse as a collection in some well known cases.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler this is ready for review.

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson @cston ptal.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:885]: >using var _ = this.GetDisposableResetPoint(resetOnDispose: true); [](http://example.com/codeflow?start=12&length=65)

I am not going to block on this, but I would prefer if this was a `using` statement with a block. The chosen style has a big negative impact on code readability and, as a result, on code maintainability as well, in my opinion.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:890]: >is always an attribute. [](http://example.com/codeflow?start=22&length=23)

Perhaps not "always", but "very likely"?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:900]: >GetLiteralExpression [](http://example.com/codeflow?start=40&length=20)

Consider using ```IsLiteralExpression``` helper instead. It matches our intent exactly.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:933]: >IsAttributeTarget [](http://example.com/codeflow?start=21&length=17)

I find the name for this helper extremely confusing/misrepresenting. The method doesn't even check if we are right after a`[` token. Consider using the name that accurately reflects what the method checks for. For example, "IsSomeWordFollowedByColon".

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11983]: > return false; [](http://example.com/codeflow?start=19&length=14)

Do we have a test covering this code path?

**Comment by AlekseyTs** [comment]
Done with review pass (commit 17)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:885]: I find it readable.  But if you prefer a block, i'll move to that.  

> In general, we care more about clarity of the code vs. brevity and lack of extra indentation.

I care about that as well.  But i find this form much more readable.  I don't need to figure out where it ends.  I know where it ends (since it extends to the end of the containing block.  So i can instantly know that any return within it resets back to a correct place.  A using with a block has to then be examined to ensure nothing happens outside that block.  This was a core benefit to this form in the first place (esp. as that's the common case for usings and try/finalys).

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: switched.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: updated.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Ah.  I missed that that exists.  THanks.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:933]: So, this follows the naming we have all over the parser.  The `Is` methods are very often for asking if we're on a particular construct, given that teh caller sensibly can ask that sort of question.  Given that both callers are effectively the attribute checking/parsing code, asking if hte position we're at is an attribute target is sensible (and fits the rest of the parser).

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: > I know where it ends (since it extends to the end of the containing block.  

You wrote the code, of course you know that now.

One of the problems is that we don't know where it begins, we don't even see that there is a `using` there, because it looks very much as a regular local declaration. We want things to be explicit, we want them to be in our face. If there is a context, we want it to be obvious.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: > You wrote the code, of course you know that now.

I know that now, and in the future.  And that's *very much* the semantics i want.  :)

With a block-style using, we've def seen people accidentally add the code *after* teh block.  This requires *much* more extensive review to ensure proper semantics.  With variable-style-using, i don't have to worry about such accidents.  This is very beneficial in my eyes or readability and maintainability.  YMMV though.  I've made the change since i don't feel strongly about it, and i'm happy to have this code match your style.

I'm just pointing out that i acutally prefer the other form not for style, but because i have found it leads to more correct code overall.

>  We want things to be explicit, we want them to be in our face.

It is explicit.  There is nothing implicit here.  As to `in our face`, that's going to be up to the individual.  I personally find `using` to be so `in your face` as to be unavoidable to miss.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: No.  An affectation from a previous path.  I've removed it.  Thanks!

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 21), assuming CI is passing

**Comment by CyrusNajmabadi** [comment]
@333fred ptal :)

**Comment by CyrusNajmabadi** [comment]
@cston  @RikkiGibson ptal

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:903]: Is this basically an error recovery path? I thought if `[` is not followed by an identifier or attribute target it is bound to be invalid as an attribute.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:903]: so this is error recovery, but attribute-parsing may still be prefered.  Consider something as simple as:

```c#
[
class C
```

We could parse this as a collection-expr as a top level statement, prior to a class.... or we can parse it as an attribute.

Right now we'll continue parsing that as an attribute, which is good here.  We can certain refine the above if we want.  But i really want to ensure large numeric arrays get better, without potentially regressing cases like this.

**Pull request merged**
